### PR TITLE
Implement Pydantic V2 protocol

### PR DESCRIPTION
**Pull request by Kludex**
- [X] closes #53999
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

1. What's the best file to add a test for this change? 😅
2. Should I implement the one for `pd.DataFrame` as well on this PR?

### CODE REVIEW

**Comment by MarcoGorelli** [comment]
thanks for your PR - could you show an example of how users would use this?

As in:
- how they can use this now
- how this PR changes/improves that

As for tests, maybe `test_downstream.py`?

File: environment.yml [hunk 1]
```diff
@@ -45,6 +45,7 @@ dependencies:
   - py
   - psycopg2>=2.9.3
   - pyarrow>=7.0.0
+  - pydantic>=2.0
```
Comment (MarcoGorelli) [inline environment.yml:48]: there's a few more files where this needs to be added

File: pandas/core/series.py [hunk 2]
```diff
@@ -6215,3 +6221,15 @@ def cumsum(self, axis: Axis | None = None, skipna: bool = True, *args, **kwargs)
     @doc(make_doc("cumprod", 1))
     def cumprod(self, axis: Axis | None = None, skipna: bool = True, *args, **kwargs):
         return NDFrame.cumprod(self, axis, skipna, *args, **kwargs)
+
+    @classmethod
+    def __get_pydantic_core_schema__(cls, source_type: Type[Any], handler: GetCoreSchemaHandler) -> CoreSchema:
+        pyd_core = cast("pydantic_core", import_optional_dependency("pydantic_core"))
```
Comment (MarcoGorelli) [inline pandas/core/series.py:6227]: why are you casting to `pydantic_core`? is that a type?

File: pandas/tests/test_downstream.py [hunk 3]
```diff
@@ -274,3 +274,23 @@ def __radd__(self, other):
 
     assert right.__add__(left) is NotImplemented
     assert right + left is left
+
+
+@td.skip_if_no("pydantic")
+@td.skip_if_no("pydantic_core")
+@pytest.mark.parametrize("series", [
+    Series([1, 2, 3]),
+    Series(np.array([1, 2, 3])),
```
Comment (MarcoGorelli) [inline pandas/tests/test_downstream.py:283]: these two are the same

File: pandas/core/series.py [hunk 4]
```diff
@@ -6215,3 +6221,15 @@ def cumsum(self, axis: Axis | None = None, skipna: bool = True, *args, **kwargs)
...
+        core_schema = pyd_core.core_schema
+
+        return core_schema.union_schema(
+            [
+                core_schema.is_instance_schema(Series),
+                core_schema.no_info_plain_validator_function(Series)
+            ]
+        )
```
Comment (MarcoGorelli) [inline pandas/core/series.py:6235]: could you explain what these do please?

File: pandas/tests/test_downstream.py [hunk 5]
```diff
@@ -274,3 +274,23 @@ def __radd__(self, other):
...
+def test_pydantic_protocol(series: Series) -> None:
+    from pydantic import BaseModel
+
+    class Model(BaseModel):
+        series: Series
+
+
+    model = Model(series=series)
+    assert model.model_dump() == {}
+    assert model.model_json_schema() == {}
```
Comment (MarcoGorelli) [inline pandas/tests/test_downstream.py:296]: could you explain these two lines please?

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
generally favourable to the idea

can you run the code checks as well please? check the contributing guide for how

File: pandas/tests/test_downstream.py (hunk 3 above)
Comment (Kludex) [reply pandas/tests/test_downstream.py:283]: yeah, I actually should have used `[1, 2,3]` and `np.array[1, 2, 3]` without the `Series`. I'll fix it.

File: pandas/core/series.py (hunk 2 above)
Comment (Kludex) [reply pandas/core/series.py:6227]: `pydantic_core` is a package: https://github.com/pydantic/pydantic-core - a dependency of Pydantic.

You can use `cast()` with the module to have your IDE understand what you are doing:

image

File: pandas/tests/test_downstream.py (hunk 5 above)
Comment (MarcoGorelli) [reply pandas/tests/test_downstream.py:296]: ok I'd suggest following https://pandas.pydata.org/docs/dev/development/contributing_environment.html

**Review by mroeschke — COMMENTED** [review]
If https://github.com/pandas-dev/pandas/issues/53999#issuecomment-1625289148 is true, then adding the pydantic protocol in pandas directly isn't a good idea if the protocol could differ on the use case

**Comment by Kludex** [comment]
> If [#53999 (comment)](https://github.com/pandas-dev/pandas/issues/53999#issuecomment-1625289148) is true, then adding the pydantic protocol in pandas directly isn't a good idea if the protocol could differ on the use case

There was a miscommunication. This is what I meant: https://github.com/pandas-dev/pandas/issues/53999#issuecomment-1627346045

**Comment by Kludex** [comment]
It's active, and this is on my TODO list. 🤷‍♂️

**Pull request closed without merging**